### PR TITLE
i#7157 syscall sched: Handle static injected syscall traces in scheduler

### PR DESCRIPTION
**Pull request by abhinav92003**
Adds handling for statically-injected kernel syscall traces in the scheduler.

Ensures that quantum and voluntary context switches are delayed until after the statically-injected syscall trace. This involved fixing the bookkeeping logic which is done on the next user-space instr now.

Note that the injected kernel syscall traces do not include any scheduling-related markers such as TRACE_MARKER_TYPE_SYSCALL_UNSCHEDULE and TRACE_MARKER_TYPE_TIMESTAMP.

For now we keep status quo on the scheduler behavior of showing the post-syscall markers before the switch.

Adds a unit test for static-injected kernel syscall trace handling by the scheduler for various scenarios involving syscall sequences shorter and longer than the quantum, for system calls that do and do not cause context switches, and occurring at different offsets into the quantum.

Issue: #7157

### CODE REVIEW

File: clients/drcachesim/scheduler/scheduler_dynamic.cpp [hunk 1]
```diff
@@ -535,14 +545,21 @@ scheduler_dynamic_tmpl_t<RecordType, ReaderType>::check_for_input_switch(
...
+            } else {
+                VPRINT(this, 4,
+                       "next_record[%d]: input %d hit end of time quantum after %" PRIu64
+                       "\n",
+                       output, input->index, input->time_spent_in_quantum);
+                preempt = true;
+                need_new_input = true;
+                input->time_spent_in_quantum = 0;
+                ++outputs_[output].stats[memtrace_stream_t::SCHED_STAT_QUANTUM_PREEMPTS];
+            }
```
Comment (derekbruening) [inline clients/drcachesim/scheduler/scheduler_dynamic.cpp:562]: Is it guaranteed there are no timestamps in kernel sequences?  Can we have a check for that?

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 2]
```diff
@@ -1354,6 +1354,183 @@ test_synthetic()
...
+            }
+            // Test a syscall sequence starting at each offset within a quantum
+            // of instrs.
+            if (i <= QUANTUM_DURATION && i == j) {
+                inputs[i].push_back(make_marker(TRACE_MARKER_TYPE_SYSCALL, 84));
+                inputs[i].push_back(
+                    make_marker(TRACE_MARKER_TYPE_SYSCALL_TRACE_START, 84));
+                for (int k = 0; k <= QUANTUM_DURATION; ++k)
+                    inputs[i].push_back(make_instr(KERNEL_CODE_OFFSET + k));
+                inputs[i].push_back(make_marker(TRACE_MARKER_TYPE_SYSCALL_TRACE_END, 84));
```
Comment (derekbruening) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:1395]: 84 is repeated a bunch of times: a named constant would be good.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 3]
```diff
@@ -1354,6 +1354,183 @@ test_synthetic()
...
+
+    // Hardcoding here for the 2 outputs and 7 inputs.
+    // We make assumptions on the scheduler's initial runqueue assignment
+    // being round-robin, resulting in 4 on core0 (odd parity letters) and 3 on
+    // core1 (even parity letters).
+    // The dots are markers and thread exits.
+    // A has a voluntary switch at its 6th instr (1st in that scheduling). Its
+    // CPU usage persists to its next scheduling which has only 2 letters.
+    // B has a voluntary switch at its 2nd instr, but it doesn't take because a
+    // syscall sequence starts just then.
```
Comment (derekbruening) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:1414]: Optional: visualize the syscall vs regular somehow: lower-case letters for syscall?

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 4]
```diff
@@ -1354,6 +1354,183 @@ test_synthetic()
...
+    // The syscall sequence consists of 4 instrs which is greater than the
+    // #instr quanta.
+    // Total instrs in A, B, C, and D are 9 + 4 == 13, others have just 9.
+
+    // Hardcoding here for the 2 outputs and 7 inputs.
+    // We make assumptions on the scheduler's initial runqueue assignment
+    // being round-robin, resulting in 4 on core0 (odd parity letters) and 3 on
+    // core1 (even parity letters).
+    // The dots are markers and thread exits.
+    // A has a voluntary switch at its 6th instr (1st in that scheduling). Its
```
Comment (derekbruening) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:1411]: After the 6th instr is here:  `A..AAAA...GA....GA^^^^AC...`.  Do you mean before the 6th instr?

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 4 above)
Comment (abhinav92003) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1411]:
```
if ((i == 0 || i == 1) && j == 1) {
```

inserts the blocking syscall at j=1.  However, for A (i==0), at j=0 we add 1 regular + 4 syscall instrs. So the blocking syscall ends up at the 6th instr ordinal.


"..A..AAAA...CCC..CCCC...EEE..GGGA....CCCEEEGGGAACCC.EEE.AAAAA."
------------------------------------^

Pointed out the 6th instr for A, after which there's a switch to C

Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1411]: I think I miscounted: the 4 A's in a row blurred into 3 or something (probably b/c all the tests generally only have 3 in a row with a quantum of 3).

Comment (abhinav92003) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1411]: Maybe independent: but the new TRACE_MARKER_TYPE_SYSCALL added doesn't have a preceding instr... need to fix that in the test case.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 3 above)
Comment (abhinav92003) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1414]: This was very helpful. Done.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 2 above)
Comment (abhinav92003) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1395]: Done

File: clients/drcachesim/scheduler/scheduler_dynamic.cpp (hunk 1 above)
Comment (abhinav92003) [reply clients/drcachesim/scheduler/scheduler_dynamic.cpp:562]: Yes that's guaranteed.  Added a check in process_marker.

File: clients/drcachesim/scheduler/scheduler_dynamic.cpp [hunk 5]
```diff
@@ -506,18 +508,26 @@ scheduler_dynamic_tmpl_t<RecordType, ReaderType>::check_for_input_switch(
...
-            // input has the oldest timestamp, prioritizing context switches
-            // over timestamp ordering.
-            VPRINT(this, 4, "next_record[%d]: input %d hit end of instr quantum\n",
-                   output, input->index);
-            preempt = true;
-            need_new_input = true;
-            input->instrs_in_quantum = 0;
-            ++outputs_[output].stats[memtrace_stream_t::SCHED_STAT_QUANTUM_PREEMPTS];
+            if (outputs_[output].in_syscall_code) {
+                VPRINT(this, 4,
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/scheduler/scheduler_dynamic.cpp:515]: Hmm.. that could get noisy. Made it level 5 for now with an XXX to print only once.

File: clients/drcachesim/scheduler/scheduler_dynamic.cpp [hunk 6]
```diff
@@ -506,18 +508,26 @@ scheduler_dynamic_tmpl_t<RecordType, ReaderType>::check_for_input_switch(
...
+                       "next_record[%d]: input %d delaying context switch "
+                       "after end of instr quantum due to syscall trace\n",
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/scheduler/scheduler_dynamic.cpp:517]: Augmented quantum_duration_instrs docs.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 4 above)
Comment (abhinav92003) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1411]: Made the test check more conditions, and added more detailed narration to make it clearer.

Comment (abhinav92003) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp]: Made the fallback check a bit stronger for the #instr quantum sequence.

**Pull request merged**